cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -336,30 +336,30 @@ std::pair<VersionedItem, FrameAndDescriptor> LocalVersionedEngine::read_datafram
     return std::make_pair(version.value_or(VersionedItem{}), std::move(frame_and_descriptor));
 }
 
-std::pair<VersionedItem, std::optional<google::protobuf::Any>> LocalVersionedEngine::read_descriptor_version_internal(
+std::pair<VersionedItem, std::optional<google::protobuf::Any>> LocalVersionedEngine::read_descriptor_internal(
         const StreamId& stream_id,
         const VersionQuery& version_query
     ) {
     ARCTICDB_SAMPLE(ReadDescriptor, 0)
-
-    auto version = get_version_to_read(stream_id, version_query);
-    missing_data::check<ErrorCode::E_NO_SUCH_VERSION>(static_cast<bool>(version),
+    auto metadata = read_metadata_internal(stream_id, version_query);
+    missing_data::check<ErrorCode::E_NO_SUCH_VERSION>(static_cast<bool>(metadata.first.has_value()),
         "Unable to retrieve descriptor data. {}@{}: version not found", stream_id, version_query);
-
-    auto metadata_proto = store()->read_metadata(version->key_).get().second;
-    return std::pair{version.value(), metadata_proto};
+    VersionedItem version{std::move(to_atom(*metadata.first))};
+    return std::pair{version, metadata.second};
 }
 
 std::vector<std::pair<VersionedItem, std::optional<google::protobuf::Any>>> LocalVersionedEngine::batch_read_descriptor_internal(
         const std::vector<StreamId>& stream_ids,
         const std::vector<VersionQuery>& version_queries) {
-    std::vector<folly::Future<std::pair<VersionedItem, std::optional<google::protobuf::Any>>>> results_fut;
-    for (size_t idx=0; idx < stream_ids.size(); idx++) {
-        auto version_query = version_queries.size() > idx ? version_queries[idx] : VersionQuery{};
-        results_fut.push_back(read_descriptor_version_internal(stream_ids[idx],
-                                                              version_query));
+    auto versions_fut = batch_get_versions(store(), version_map(), stream_ids, version_queries);
+    std::vector<folly::Future<std::pair<VersionedItem, std::optional<google::protobuf::Any>>>> fut_vec;
+    for (auto& version: versions_fut){
+        fut_vec.push_back(get_metadata_async(std::move(version))
+        .thenValue([](auto&& meta){
+            return std::make_pair(VersionedItem{std::move(to_atom(meta.first.value()))}, std::move(meta.second));
```

Comment:
This changes the behaviour from a useful `E_NO_SUCH_VERSION` exception to a cryptic `std::bad_optional_access` if one of the versions doesn't exist.
It would be less error prone to have a `read_descriptor_async` method as we do for metadata, so that it is impossible for these discrepancies between batch and non-batch methods to creep in.

Reply:
agree. Done.